test-engine/src/main/java/com/aws/jverify/testengine/JVerifyTestEngine.java

```
@@ -0,0 +1,250 @@
+import com.aws.jverify.verifier.VerifierOptions;
+import com.google.auto.service.AutoService;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Assumptions;
+import org.junit.platform.commons.support.ReflectionSupport;
+import org.junit.platform.engine.DiscoverySelector;
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.ExecutionRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.discovery.ClassSelector;
+import org.junit.platform.engine.discovery.ClasspathRootSelector;
+import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor;
+import org.junit.platform.engine.support.descriptor.EngineDescriptor;
+import org.junit.platform.engine.support.hierarchical.EngineExecutionContext;
+import org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine;
+import org.junit.platform.engine.support.hierarchical.Node;
+
+import javax.tools.Diagnostic;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@AutoService(TestEngine.class)
+public class JVerifyTestEngine extends HierarchicalTestEngine<EngineExecutionContext> {
+    static final Logger LOGGER = Logger.getLogger(JVerifyTestEngine.class.getName());
+
+    @Override
+    public String getId() {
+        return "jverify-test-engine";
+    }
+
+    @Override
+    public TestDescriptor discover(EngineDiscoveryRequest request, UniqueId uniqueId) {
+        var descriptor = new EngineDescriptor(uniqueId, "JVerify Tests");
+        request.getSelectorsByType(DiscoverySelector.class).stream()
```

Since you're having to look at the source directory below anyway, why not discover all source files in that directory and then filter them to the files that match the discovery selectors? That will solve the problem of non-public classes not working.

---

I'm not sure what problem you're referring to - by "non-public classes" do you mean classes named differently that the source file in which they're declared? If so, I don't think this will solve the issue unless we re-parse the source files during test discovery in search of the right `package ...` + `class ...` combination, and I don't love that idea.

---

Yes sorry, classes with the "wrong" class name. Currently they are silently skipped. Given other discussions I'm happy if they just lead to an error and we don't support them.

---

They're not silently skipped, in the sense that if you add `@JVerifyTest` then you'll get an error when running the test as a class or when running a test scope that includes the class (e.g. the file itself, the package, or the entire module/project), and if you don't add the annotation then you won't be able to run the test at all and the linter will flag the class as unused.

---

It was silently skipped in the version I tried but now that you've made the warning an error it fails, so I'm happy. :)